File: sdk/python/dogdata/resources/holders.py
from __future__ import annotations

from typing import Any, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from dogdata.client import DogData


class HoldersResource:
    def __init__(self, client: DogData):
        self._client = client

    def list(self, page: int = 1, limit: int = 100) -> dict[str, Any]:
        return self._client.request("/api/dog-rune/holders", {
            "page": page,
            "limit": limit,
        })

    def get(self, address: str) -> dict[str, Any]:
        return self._client.request("/api/dog-rune/holders", {
            "address": address,
        })

    def count(self) -> dict[str, Any]:
        return self._client.request("/api/dog-rune/holders", {
            "count": True,
        })

    def top(self, limit: int = 100) -> list[dict[str, Any]]:
        result = self.list(page=1, limit=limit)
        return result.get("data", [])

File: sdk/typescript/src/resources/holders.ts
import type { DogData } from '../client.js';
import type { PaginationParams, PaginatedResponse, Holder, HolderDetail } from '../types.js';

export class HoldersResource {
  constructor(private client: DogData) {}

  async list(params?: PaginationParams): Promise<PaginatedResponse<Holder>> {
    return this.client.request<PaginatedResponse<Holder>>('/api/dog-rune/holders', {
      page: params?.page,
      limit: params?.limit,
    });
  }

  async get(address: string): Promise<HolderDetail> {
    return this.client.request<HolderDetail>(`/api/dog-rune/holders`, {
      address,
    });
  }

  async count(): Promise<{ total: number }> {
    return this.client.request<{ total: number }>('/api/dog-rune/holders', {
      count: true,
    });
  }

  async top(limit: number = 100): Promise<Holder[]> {
    const result = await this.list({ page: 1, limit });
    return result.data;
  }
}
